Merkle payments Infrastructure (without CLI)

grumbach opened the pull request:
# Merkle Payments Infrastructure

Adds merkle batch payment infrastructure to enable cheaper gas fees by batching on-chain payment for data uploads. This PR does not include the CLI integration which is coming soon...

## Key Changes

**Protocol & API**
- New `Query::GetMerkleCandidateQuote` for requesting merkle payment quotes from nodes
- `RecordKind::DataWithMerklePayment` variant for merkle-paid records
- Client methods for merkle-based uploads: `upload_with_merkle_payments`, `put_with_merkle_payment`

**Node**
- Merkle proof verification in `PutValidation` for all data types
- Node-side quote generation with cryptographic signatures binding PeerId to RewardsAddress
- Payment expiration validation

**EVM Integration**
- Complete `ant-evm::merkle_payments` module with merkle tree construction and verification
- Support for up to 1024 leaves per merkle tree
- Smart contract integration for on-chain payment verification

This infrastructure enables batch payment verification where multiple chunks can be paid for with a single on-chain transaction using merkle proofs, significantly reducing gas costs compared to per-chunk payments.

grumbach committed:
feat: merkle tree protocol

grumbach committed:
fix: rs merkle import

grumbach committed:
feat: ant-evm side payment impl with mocked smart contract

grumbach committed:
fix: use XorName hashing instead of blake2

grumbach committed:
feat: protocol integration

grumbach committed:
feat: node integration

grumbach committed:
fix: various issues in node integration

grumbach committed:
feat: client integration

grumbach committed:
feat: client merkle payment integration

grumbach committed:
feat: CLI integration

grumbach committed:
feat: add cost info to merkle payment APIs

grumbach committed:
feat: smart contract integration

grumbach committed:
fix: merkle payment timestamp clock skew support node side

grumbach committed:
fix: make quoting more resilient to bad nodes

grumbach committed:
fix: sybil closest peer deduplication in merkle payments

grumbach committed:
feat: add missing contract abi json

grumbach committed:
feat: improved printing and reduced constants

grumbach committed:
feat: use stream_encrypt instead of archaic streaming_encrypt_from_file

grumbach committed:
feat: better print feedback for merkle payments

grumbach committed:
feat: add retries to the smart contract event waiting

grumbach committed:
chore: various improvements

grumbach committed:
fix: critical address mismatch on nodes topology checks

grumbach committed:
fix: add more retries to merkle smart contract event getting

grumbach committed:
feat: reduce memory load of merkle uploads

grumbach committed:
feat: cheaper uploads and smart contract generation modules

grumbach committed:
feat: simplify chunk batching in merkle uploads

grumbach committed:
feat: add ci test for merkle

grumbach committed:
chore: various fixes

grumbach committed:
feat: more robust evm fetching

grumbach committed:
feat: distance in topology checks

grumbach committed:
feat: support upload of large files in 4GB batches

mickvandijke committed:
refactor(evmlib): modularize `merkle_payment_vault` and update contract

mickvandijke committed:
refactor(evmlib): add retry for `PoolHash` get

grumbach committed:
feat: refactored pay, upload, then next batch flow in merkle payments

grumbach committed:
chore: set max depth to 10

mickvandijke committed:
refactor(evmlib): improve merkle payment vault error handling

grumbach committed:
chore: usability improvements and code quality

grumbach committed:
fix: critical stream exhausted bug

grumbach committed:
chore: unused vars without loud flag

grumbach committed:
feat: add merkle sepolia contract hardcoded

mickvandijke committed:
refactor(evmlib): reduce MAX_MERKLE_DEPTH to 8

jacderida committed:
chore: support `--merkle-payments-address` argument

The `antctl add` command needs to emit the `--merkle-payments-address` argument in the service
definition so Merkle Payments can be supported for testnet deployments.

The commit also includes a `cargo fmt` fixup.

grumbach committed:
chore: merge conflicts and compile issues

grumbach committed:
fix: critical network task handler bug

mickvandijke committed:
feat: add detailed logging for topology verification failures

mickvandijke committed:
refactor: simplify node side topology failure log

mickvandijke committed:
feat(ant-node): use same get_closest_peers method as client for topology verification

mickvandijke committed:
chore: fix clippy errors

mickvandijke committed:
fix: docs error

mickvandijke committed:
refactor(ant-node): rename `get_closest_peers_verified` to `get_closest_peers_with_majority_knowledge` for clarity

grumbach committed:
fix: avoid cloning xorname list

grumbach committed:
fix: remove abusive logging which clogged memory

mickvandijke committed:
feat(evmlib): add `ARBITRUM_ONE_MERKLE_PAYMENTS_ADDRESS` const

grumbach committed:
feat: disable wip cli work for node release

grumbach committed:
chore: clippy

grumbach committed:
chore: cleanup merkle cli tests for node release

grumbach commented on ant-evm/Cargo.toml:
```diff
@@ -15,12 +15,14 @@ test-utils = []
 rand = { version = "~0.8.5", features = ["small_rng"] }
 ring = "0.17.12"
 rmp-serde = "1.1.1"
+rs_merkle = { git = "https://github.com/maidsafe/rs-merkle.git", branch = "master" }
```
@jacderida do we need to release this properly for a release?

grumbach committed:
chore: remove udep

grumbach committed:
chore: properly feature gated mock disk smart contract

maqi commented on ant-protocol/src/storage/header.rs:
```diff
@@ -93,23 +102,28 @@ impl<'de> Deserialize<'de> for RecordKind {
+            "DataOnly" => Self::DataOnly(data_type),
+            "DataWithPayment" => Self::DataWithPayment(data_type),
+            "DataWithMerklePayment" => Self::DataWithMerklePayment(data_type),
+            _ => unreachable!(),
```
define a helper enum to avoid `unreachable`, which is not perferred in the production code

maqi commented on ant-protocol/src/peer_id_serde.rs:
```diff
@@ -0,0 +1,31 @@
+
+//! Helper module for serializing/deserializing `Vec<PeerId>`
+
+use libp2p::PeerId;
```
we use NetworkAddress for PeerID messaging between clients/nodes.

maqi commented on ant-protocol/src/messages/query.rs:
```diff
@@ -84,6 +84,24 @@ pub enum Query {
+    /// Node signs its current state (metrics + reward address) with the
+    /// provided timestamp, creating a commitment that cryptographically
+    /// proves the binding between PeerId and RewardsAddress.
+    GetMerkleCandidateQuote {
```
Using a this new message not only make the merkle quotes to be outstanded from normal quotes, which could have potential backward compatible issue or being vlunerable to be identified out and gaming
But also lost the `storage challenge` feature which could be useful in near future.

Hence, I'd suggest use the same Quote req/rsp.

maqi reviewed (approved):
Though the new `GetMerkleCandidateQuote` messging no longer supports the storage check, the same check can still be carried out by having an additional round of normal `Quote` before/after the Merkle quoting.

It may affect the performance, but given there is now only `commitment pool` nodes to be queried, which is must less than the previous `nodes close to ALL chunks`, the extra round of quoting could be acceptable.

jacderida committed:
chore: use new `ant-merkle` crate

This replaces the use of the forked branch with our own published version of the crate.

jacderida committed:
Merge branch 'main' into merkle_payments_no_cli

jacderida committed:
chore: clippy fixups for new rust version
